cirq/contrib/quirk/cells/parse.py

```diff
@@ -12,121 +12,376 @@
+def _merge_scientific_float_tokens(tokens: Iterable[str]) -> List[str]:
+    tokens = list(tokens)
+    i = 0
+    while 'e' in tokens[i + 1:]:
+        i = tokens.index('e', i + 1)
+        s = i - 1
+        e = i + 1
+        if not re.match('[0-9]', str(tokens[s])):
+            continue
+        if re.match('[+-]', str(tokens[e])):
+            e += 1
+        if re.match('[0-9]', str(tokens[e])):
+            e += 1
+            tokens[s:e] = [''.join(tokens[s:e])]
+            i -= 1
+    return tokens
 
-    try:
-        result = _parse_scalar(formula, t=sympy.Symbol('t'))
-    except Exception as ex:
-        raise ValueError(
-            f"Failed to parse the gate formula {repr(formula)}.\n"
-            "This is likely due to a bug where cirq fails to exactly emulate "
-            "Quirk's parsing. Please report it.") from ex
-    if isinstance(result, (float, int, complex)) or not result.free_symbols:
-        result = complex(result)
-        if abs(np.imag(result)) > 1e-8:
+
+T = TypeVar('T')
+
+
+def _segment_by(seq: Iterable[T], *,
+                key: Callable[[T], Any]) -> Iterator[List[T]]:
+    group: List[T] = []
+    last_key = None
+    for item in seq:
+        item_key = key(item)
+        if len(group) and item_key != last_key:
+            yield group
+            group = []
+        group.append(item)
+        last_key = item_key
+    if len(group) > 0:
+        yield group
+
+
+def _tokenize(text: str) -> List[str]:
+    result = []
+    for part in re.split(r'\s', text.lower()):
+
+        def classify(e: str) -> str:
+            if e.strip() == '':
+                return " "
+            if re.match(r'[.0-9]', e):
+                return "#"
+            if re.match(r'[_a-z]', e):
+                return "a"
+            return np.nan  # Always split.
+
+        for group in _segment_by(part, key=classify):
+            result.append(''.join(group))
+
+    return _merge_scientific_float_tokens(g for g in result if g.strip())
+
+
+_ResolvedToken = Union[sympy.Basic, int, float, complex]
+
+
+class _CustomQuirkOperationToken:
+
+    def __init__(
+            self,
+            unary_action: Optional[Callable[[_ResolvedToken], _ResolvedToken]],
+            binary_action: Optional[
+                Callable[[_ResolvedToken, _ResolvedToken], _ResolvedToken]],
+            priority: float):
+        self.unary_action = unary_action
+        self.binary_action = binary_action
+        self.priority = priority
+
+
+class _HangingNode:
+
+    def __init__(
+            self,
+            func: Callable[[_ResolvedToken, _ResolvedToken], _ResolvedToken],
+            weight: float):
+        self.func = func
+        self.weight = weight
+
+
+_HangingToken = Union[_ResolvedToken, str, _CustomQuirkOperationToken]
+
+
+def _translate_token(token_id: str,
+                     token_map: Mapping[str, _HangingToken]) -> _HangingToken:
+    if re.match(r'[0-9]+(\.[0-9]+)?', token_id):
```

Comment:
Note that `re.match` matches the regex *at the beginning* of text, i.e. `re.match('a', 'ab')` is considered a match (see also [documentation](https://docs.python.org/3/library/re.html)). Therefore, for defense-in-depth I'd append `$` to the regex to force exact match:

```python
if re.match(r'[0-9]+(\.[0-9]+)?$', token_id):
```

This way someone reading the code doesn't have to wonder whether `float(token_id)` raises `ValueError` or not.

Reply:
Acknowledged, but I'd prefer to keep it closer to the Quirk implementation.